x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -451,17 +453,47 @@ public void writeTo(StreamOutput out) throws IOException {
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
-        builder.field(ID.getPreferredName(), id);
-        builder.field(Job.ID.getPreferredName(), jobId);
-        if (params.paramAsBoolean(ToXContentParams.FOR_INTERNAL_STORAGE, false)) {
-            builder.field(CONFIG_TYPE.getPreferredName(), TYPE);
+        if (params.paramAsBoolean(FOR_EXPORT, false) == false) {
+            builder.field(ID.getPreferredName(), id);
+            // We don't include the job_id in export as we assume the PUT will be referring to a new job as well
+            builder.field(Job.ID.getPreferredName(), jobId);
+            if (params.paramAsBoolean(ToXContentParams.FOR_INTERNAL_STORAGE, false)) {
+                builder.field(CONFIG_TYPE.getPreferredName(), TYPE);
+            }
+            if (headers.isEmpty() == false && params.paramAsBoolean(ToXContentParams.FOR_INTERNAL_STORAGE, false)) {
+                builder.field(HEADERS.getPreferredName(), headers);
+            }
+            builder.field(QUERY_DELAY.getPreferredName(), queryDelay.getStringRep());
+            builder.field(QUERY.getPreferredName(), queryProvider.getQuery());
+            if (chunkingConfig != null) {
+                builder.field(CHUNKING_CONFIG.getPreferredName(), chunkingConfig);
+            }
+            builder.startObject(INDICES_OPTIONS.getPreferredName());
+            indicesOptions.toXContent(builder, params);
+            builder.endObject();
+        } else { // Don't include random defaults or unnecessary defauls in export
```

Comment:
So this `else` block is not for making the retrieved config "puttable" but it does something more (clears defaults).
I'm wondering what happens when we ever change the code for generating the default. Then the config is indexed with default `D1`, then in the subsequent version we change the default to be `D2` and then we retrieve the config for export and the default is not cleared (as it was the default in the past but no longer is). Is that a good behavior?

Reply:
@przemekwitek a related question.

What if the user didn't set their own `chunking_config` but when they cloned the datafeed, they DID change the aggregation information (maybe the `date_histogram` bucket size). Then when they try to `PUT`, the chunking config is now illegal. 

Should we allow this behavior?

Reply:
Tough question. I guess it's better to stick to the promise that we are able to `PUT` the config obtained via `GET` with `for_export`...